src/core/unit.h

```diff
@@ -106,6 +106,62 @@ typedef union UnitDependencyInfo {
         } _packed_;
 } UnitDependencyInfo;
 
+/* Store information about why a unit was activated.
+ * We start with trigger units (.path/.timer), eventually it will be expanded to include more metadata. */
+typedef struct ActivationEventInfo {
+        UnitType trigger_unit_type;
+        unsigned n_ref;
+        char *trigger_unit_name;
+} ActivationEventInfo;
+
+/* For casting an activation event into the various unit-specific types */
+#define DEFINE_EVENT_INFO_CAST(UPPERCASE, MixedCase, UNIT_TYPE)                 \
+        static inline MixedCase* UPPERCASE(ActivationEventInfo *a) {            \
+                if (_unlikely_(!a || a->trigger_unit_type != UNIT_##UNIT_TYPE)) \
+                        return NULL;                                            \
+                                                                                \
+                return (MixedCase*) a;                                          \
+        }
+
+/* For casting the various unit types into a unit */
+#define ACTIVATION_EVENT_INFO(u)                                      \
+        ({                                                            \
+                typeof(u) _u_ = (u);                                  \
+                ActivationEventInfo *_w_ = _u_ ? &(_u_)->meta : NULL; \
+                _w_;                                                  \
+        })
+
+ActivationEventInfo *activation_event_info_new(Unit *trigger_unit);
+ActivationEventInfo *activation_event_info_ref(ActivationEventInfo *p);
+ActivationEventInfo *activation_event_info_unref(ActivationEventInfo *p);
+void activation_event_info_serialize(ActivationEventInfo *p, FILE *f);
+int activation_event_info_deserialize(const char *key, const char *value, ActivationEventInfo **info);
+DEFINE_TRIVIAL_CLEANUP_FUNC(ActivationEventInfo*, activation_event_info_unref);
+
+typedef struct ActivationEventInfoVTable {
+        /* How much memory does an object of this activation type need */
+        size_t object_size;
+
+        /* This should reset all type-specific variables. This should not allocate memory, and is called
+         * with zero-initialized data. It should hence only initialize variables that need to be set != 0. */
+        void (*init)(ActivationEventInfo *a, Unit *trigger_unit);
+
+        /* This should free all type-specific variables. It should be idempotent. */
+        void (*done)(ActivationEventInfo *a);
+
+        /* This should serialize all type-specific variables. */
+        void (*serialize)(ActivationEventInfo *a, FILE *f);
+
+        /* This should deserialize all type-specific variables, one at a time. */
+        int (*deserialize)(const char *key, const char *value, ActivationEventInfo **a);
+} ActivationEventInfoVTable;
```

Comment:
oh, i would not bother with a vtable really, it's not that we need 100 fields or so in the info `ActivationEventInfo` structure. i'd just do separate fields in it that migh tbe unit type specific. let's keep this simple...

Reply:
I've just added a couple more, and with the plans you outlined to extend it further there will be even more, most likely. So I'd like to keep it. It looks much more elegant and allows to extend to other unit types with a lot less churn.